includes/admin/upgrades/upgrade-functions.php

```diff
@@ -1328,498 +1328,230 @@ function edd_include_file_download_log_migration_batch_processor( $class = '' )
-				</span>
-				</form>
-			</div><!-- .inside -->
-		</div><!-- .postbox -->
+					<span class="edd-migration-percentage edd-hidden">
+						&ndash;
+						<span class="edd-migration-percentage-value">0</span>%
+					</span>
+				</li>
+			<?php endforeach; ?>
+		</ul>
 	</div>
-	<?php $step++; ?>
 
-	<div class="metabox-holder">
-		<div class="postbox">
-			<h2 class="hndle">
-				<span>
-					<?php echo esc_html( sprintf( /* translators: %d Migration step count */ __( 'Step %d: Upgrade Customer Notes', 'easy-digital-downloads' ), $step ) ); ?>
-				</span>
-				<span class="dashicons dashicons-yes"></span>
-			</h2>
-			<div class="inside migrate-customer-notes-control">
-				<p>
-					<?php esc_html_e( 'This will migrate all customer notes from the WordPress posts table to custom database tables for improved performance and reliability.', 'easy-digital-downloads' ); ?>
-				</p>
-				<form method="post" id="edd-migrate-customer-notes-form" class="edd-export-form edd-import-export-form">
-				<span class="step-instructions-wrapper">
-					<?php wp_nonce_field( 'edd_ajax_export', 'edd_ajax_export' ); ?>
-
-					<?php if ( ! $customer_notes_removal_complete ) : ?>
-						<span class="edd-migration allowed" style="<?php echo $migration_complete ? '' : 'display: none'; ?>">
-							<button type="submit" id="migrate-customer-notes-submit" class="button button-primary"><?php esc_attr_e( 'Upgrade Database', 'easy-digital-downloads' ); ?></button>
-						</span>
-
-						<span class="edd-migration unavailable" style="<?php echo ! $migration_complete ? '' : 'display: none'; ?>">
-							<button type="submit" disabled id="migrate-customer-notes-submit" class="button button-secondary"><?php esc_attr_e( 'Upgrade Database', 'easy-digital-downloads' ); ?></button>
-							&mdash;
-								 <span class="message initial"><?php esc_html_e( 'Please complete the previous steps before updating customer notes.', 'easy-digital-downloads' ); ?></span>
-								 <span class="message complete"><?php esc_html_e( 'Data migration complete.', 'easy-digital-downloads' ); ?></span>
-						</span>
-					<?php else: ?>
-						<button type="submit" disabled id="migrate-customer-notes-submit" class="button button-secondary"><?php esc_attr_e( 'Upgrade Database', 'easy-digital-downloads' ); ?></button>
-						&mdash; <?php esc_html_e( 'Legacy data has already been removed, migration is not possible at this time.', 'easy-digital-downloads' ); ?>
-					<?php endif; ?>
-
-					<input type="hidden" name="edd-export-class" value="EDD\Admin\Upgrades\v3\Customer_Notes" />
-				</span>
-				</form>
-			</div><!-- .inside -->
-		</div><!-- .postbox -->
+	<div id="edd-v3-migration-complete" <?php echo ! $can_remove_legacy_data ? 'class="edd-hidden"' : ''; ?>>
+		<p>
+			<?php esc_html_e( 'The data migration has been successfully completed. You may now leave this page or proceed to remove legacy data below.', 'easy-digital-downloads' ); ?>
```

Comment:
I'm not sure what I would prefer, but I feel like I want this to be worded slightly differently. Trying to think of what I want without it being redundant with the explanation below.

Reply:
Is there a specific part that's bothering you? Just so I can help think about what to tinker with.

Reply:
> Also wondering--and being lazy because I'm on a different branch and can't run the migration again right this moment--but if the migration part is complete, but the legacy data is still present, does the admin notice outside of the migration screen still persist? Mainly I'm wanting to make sure that if someone leaves this screen and dutifully checks their site data, that they can easily get back, and know that they do _need_ to get back and remove the legacy data.

The admin notice does not persist. However, you can return back to the same page and you'll still have the option to remove legacy data.

I don't think I want to persist the admin notice, because although removing legacy data is good for performance, it's not a required step. Also, some people might understandably want to wait a few days before removing it to make sure their store is perfectly operational. We shouldn't persist a notice that long.

I'd be more inclined to add something to the Tools page or similar that helps make it easier to find the "Remove Legacy Data" page again.

Reply:
That would be fine by me. I'm concerned that if a user navigates away (for example, to let it wait a few days, which totally makes sense), they're going to have a hard time finding that screen again to remove the legacy data and never be able to find it again (and never know either).

So maybe add a notice with a link to the tools screen if there is still legacy data (or are you saying to add it as a tool, because that would work as well)? And update this message to say something like:

> The data migration has been successfully completed. You may now remove the legacy data, or you can leave this page. If you leave this page without removing the legacy data, you can use the Legacy Data Removal tool anytime in the future on the Downloads > Tools screen.

Reply:
https://github.com/easydigitaldownloads/easy-digital-downloads/pull/8202/commits/f75667172c2f89e828e8046ad18171ddcdcc6450 adds the same migration tool to the tools page. It only appears (hopefully) if you've completed all migration steps but have not yet removed legacy data.

I made it exactly the same in both places (mostly) to try to reuse as much code as possible.

Reply:
The legacy data tool does not show on the tools screen after I have used it to remove the legacy data.

However, it is showing on a "new" 3.0 site...not new, but it's not had 2.9 on it at all, and I did just reset the store data this morning. (It is a multisite setup and the other child sites are _not_ showing the tool, so ??)